ministryofjustice/laa-criminal-applications-datastore
# CRIMAP-268 Setup of job processor and SQS messaging

## Description of change
NOTE: this is an initial setup, kind of a proof of concept, including some "boilerplate code" and assumptions.

We can use this as a starting point to iterate once we know the actual requirements and also once we hook this up to the actual "submission" journey.

Some things to note:

1. I've focused a lot on this being easy to run in localhost, for instance with docker is super easy (similar to what we had with DynamoDB), and I've also written a small shell script to start/stop the job processor, taking as base one I did in the past for Sidekiq. This is more important later in k8s.

2. It will not use AWS SQS just yet on staging even if we deploy it. In fact it will not work on staging as there is nothing yet done on kubernetes side (but the app will not have any impact, it is safe to deploy this code).

3. Quite a few things to decide still, like queue being fifo or not, delays, visibility timeouts, etc. and of course the format of the message being push to MAAT. All this will come more down the line.

4. It is not wired up to the submissions, on purpose. This can come later on. See below on how to run/test this locally.

## Link to relevant ticket
https://dsdmoj.atlassian.net/browse/CRIMAP-268

## Notes for reviewer / how to test
A way to test this is manually by doing the following:

- run ElasticMQ with docker `docker-compose up elasticmq`
- run the jobs processor with: `./shoryuken.sh start`

You can then go to this URL to observe the queues (it updates in realtime, no need to refresh): http://localhost:9325

Finally, open a datastore ruby console and enqueue a job with, for instance: `ApplicationSubmittedJob.perform_later(CrimeApplication.last)`

The worker will pick this up in a few seconds, and process it, which in turn will produce a message being published into the MAAT queue.

You can also spin up the app with all the dependencies with `docker-compose up` and then open a rails terminal in the web container and create an app and enqueue it. It will work the same.

![Screenshot 2023-02-02 at 16 20 52](https://user-images.githubusercontent.com/687910/216381763-7c5d7862-9327-4784-8583-3e5b1254ccad.png)
